Bump to python3 and remove codeface dependency where possible

maxloeffler opened the pull request:
## Changes:

* Make the entire codebase compatible with `python3` using the python module `modernize`.
* Remove the dependency on a running [codeface](https://github.com/se-sic/codeface) instance by encapsulating all necessary `codeface` fragments into the codebase.
* Remove the dependency on `codeface` ID databases by adding an alternative ID manager service that operates on .csv files. The codeface ID database will still be required to perform the extraction process.

## Open Ties:

* How should I deal with the copyright header in the included `codeface` fragments?
* Due to a lack of data I could not yet test all of `codeface-extraction`'s functionality.

maxloeffler committed:
Automatically convert to python3 (using 2to3)

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Import codeface files required for extraction

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Remove dependency on codeface

maxloeffler committed:
Rename folder for codeface fragments

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Introduce CSV-based IdManager class

Using the csvIdManager, we can use all functionality but extraction
without a codeface installation. The csvIdManager then replaces the
MySQL-based IdManager from codeface.

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Adapt usage of codeface fragments in all modules

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Increase consistency with pyhton3 conventions

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -693,10 +690,10 @@ def get_id_and_update_user(user, buffer_db_ids=user_id_buffer, buffer_usernames=
-            name = username
-        mail = str(user["email"]).encode("utf-8")
+            name = username.decode("utf-8")
+        mail = user["email"]
```
I don't get this change. Can you please explain this to me @maxloeffler?
Why do we need to call `username.decode("utf-8")`? And why can we be sure that `user["email"]` and `user["name"]` are already in utf-8?
(not only here, same for jira_issue_processing)

maxloeffler replied:
To be honest, I can't be sure whether this solution is generally applicable, thats another reason why it would be nice to have some more testing data. The reason is that with the data I have, `user["name"]` is already of type `str` here. Calling encode on it turns it into `bytes` and this leads to the incorrect names with leading `b`s in the output csv file that we discussed in the last meetings.

What fortifies me in my expression that this encoding shift is correct is that directly after the lines you commented on, there is a call `get_user_string(name, mail)` in which concats `name` and `mail` into one `user_string`. For this concatenation to work as expected `name` and `mail` *must* be a `str` not `bytes`.

bockthom commented on issue_processing/jira_issue_processing.py:
```diff
@@ -611,11 +608,11 @@ def get_user_string(name, email):
-            name = str(user["username"]).encode("utf-8")
-        mail = str(user["email"]).encode("utf-8")  # empty
+            name = username.decode("utf-8")
+        mail = user["email"]
```
Please keep the "empty" comment. Email might still be empty in this case

bockthom reviewed (commented):
Thanks @maxloeffler for helping us with bumping the code to python3 and getting partly independent of Codeface.

Regarding the files taken from Codeface: Can we remove the functions that we don't need? For example, PersonInfo.py contains classes like `RelationWeight` or `RelationWeights`  or functions like `addReceiveRelation` - stuff that we don't use here. Can we get rid of all that stuff (in best case, via rebase) and only keep the stuff we need? (Only exception: For idManager and dbmanager we might need to keep everything for completeness reasons).

Regarding the copyright headers: They are entirely outdated in Codeface, nobody has ever updated them (even not me). Technically, we would need to figure out who has contributed to the lines that we take, but not sure. When we take it, it becomes part of codeface-extraction. Maybe we should add a link to the corresponding file in codeface and state that this code originates from there.

And for the dbmanager we should mark what is the original code and what is the new csv-manager you have added. 

This may need further discussion.

maxloeffler committed:
Update copyright headers

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Remove unused code artifacts

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

maxloeffler committed:
Ensure string represenation of all user data in 'get_id_and_update_user'

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

bockthom commented:
You did not add your name to the copyright header in a number of files which you have changed just to convert everything from Python2 to Python3. Technically speaking, the Python-version-related changes are changes to the file, so you should add your name to all files in which at least one line has changed.

Or do you have a different opinion on that @maxloeffler @hechtlC? Let's discuss this!

maxloeffler commented on issue_processing/issue_processing.py:
```diff
@@ -719,16 +720,16 @@ def get_user_from_id(idx, buffer_db=user_buffer):
-        user["id"] = person["id"]  # column "id"
+        user["email"] = person.getEmail()  # column "email1"
+        user["name"] = person.getName()  # column "name"
+        user["id"] = person.getID()  # column "id"
```
Very nice catch, there was indeed an inconsistency in the return types in `getPersonFromID` between the `dbIdManager` and the `csvIdManager`. Through making it consistent, I could remove another `codeface` file, namely `PersonInfo`.

maxloeffler commented on codeface_extraction/extractions.py:
```diff
@@ -765,12 +766,11 @@ def fix_name_encoding(name):
+        return str(name.decode('utf-8'))
     except LookupError:
         # Encoding not found, return string as is
         return name
```
I am unsure what Copilot wants to tell us here. It is true that the `except` cases are not generic enough to cover all types of exceptions but they cover the expected ones and when an unexpected exception gets raised, then it *should* probably fly through. Also I am not sure which code it refers to here as "unreachable".

Also this section has been like this prior to my PR.

maxloeffler commented on csv_writer/csv_writer.py:
```diff
@@ -28,7 +28,7 @@ def __encode(line):
     for column in line:
-        if type(column) is unicode:
+        if type(column) is str:
             lineres += (column.encode("utf-8"),)
```
Actually, the function is *not* being used anymore, so let me remove it.

maxloeffler committed:
Use the same return format in all implementations of 'getPersonFromDB'

Signed-off-by: Maximilian Löffler <s8maloef@stud.uni-saarland.de>

bockthom commented on issue_processing/issue_processing.py:
```diff
@@ -726,11 +726,8 @@ def get_user_from_id(idx, buffer_db=user_buffer):
-        user["name"] = person.getName()  # column "name"
-        user["id"] = person.getID()  # column "id"
+        user = idservice.getPersonFromDB(idx)
+        user["email"] = user.pop("email1")
```
Hm, I am wondering whether this returns the correct `user` dict. I guess `user` will have more information than `user` previously had, as previously it had only the fields email, name, and id. But now it has all fields that are present in the person object that is obtained form the database.

This might impact the amount of data that is being held in memory, and also could lead to problems later on when writing the data to disk if not only the relevant three fields are selected. Could you please check what happens with all the useless attributes, are they used/removed later on? (This is not a request to change this already, please check that first and reply before changing this.) Applies similarly also to jira_issue_processing.py

maxloeffler replied:
Oh thats something that I didn't even consider. While newly added users *should* never have any data fields but email, id, and name that may not be the case when using prefilled codeface databases. I cannot practically test this but [the definition](https://github.com/se-sic/codeface/blob/9eaaa767792feb7bdf4e3346e682ba4209947f9f/datamodel/codeface_schema.sql#L37-L61) of the `person` table that is [being queried](https://github.com/se-sic/codeface/blob/9eaaa767792feb7bdf4e3346e682ba4209947f9f/id_service/id_service.js#L115-L125) by the `dbIdManager` [here](https://github.com/se-sic/codeface/blob/9eaaa767792feb7bdf4e3346e682ba4209947f9f/codeface/cluster/idManager.py#L196) indicates the potential presence of more columns. With your approval, I would revert the code sequence to exclude additional columns, e.g.:

```python
person = idservice.getPersonFromDB(idx)
user = {
    "name": person["name"],
    "email": person["email1"],
    "id": person["id"]
}
```

bockthom replied:
Thanks for checking. I agree with your suggestion.

bockthom committed:
Properly initialize new config entry 'useCsv'

If useCsv is not set in the config file, set its default to False.
Without setting a default, it would always end up in a KeyError
when accessing the non-existing entry.

Signed-off-by: Thomas Bock <bockthom@cmu.edu>

bockthom committed:
Setup logger to log INFO logs and format log out

Configure the new logger to log at INFO level (otherwise nothing would be
logged in a normal run), and format the log output similarly to how it was
formatted previously in python2.

Signed-off-by: Thomas Bock <bockthom@cmu.edu>

bockthom commented:
While trying the python3 version in action for the very first time, I figured out the following two problems:

1. The newly introduced config variable `useCsv` did not have a default value. When this variable is missing in the database config file, the scripts ran into a KeyError as the `useCsv` key was not present in the config dictionary. I fixed this by setting this entry to `False` if no such key is present. This way we can run the script as usual without the need to change the config files.
2. The new logger did not log anything. Reason: The default log level of the new logger that @maxloeffler has added is `WARN`. However, we want to get the `INFO` level logs as previously for the python2 version. To do so, I added a setup function to `codeface_util/utils.py` that format the log output and defaults to `INFO` level. And I now call this setup function from all main scripts when initializing the logger.

I have already added two commits to this PR that fix these two issues.

I will continue adding commits to this PR (i.e., @maxloeffler's master branch) if I encounter any additional issues that were introduced in this PR in case they need to be fixed prior to merging this PR.

bockthom committed:
Fix encoding in 'encode_as_utf8'

Signed-off-by: Thomas Bock <bockthom@cmu.edu>
